API Server as event sink

v0d1ch opened the pull request:
This PR refactors the API server to remove persistence argument and instead pass in event source in order to make the API server just another event sink. 
In turn we are able to reduce the memory footprint of the hydra-node just by streaming events from disk instead of loading them all in memory. On top of this hydra-node does not serve history by default anymore.

Notable changes that came out of this PR:
:snowflake:  API server doesn't use persistence any more and is now just another event sink
:snowflake: `PersistenceIncremental/source` doesn't check anymore for the authorized thread since we made the implementation work across threads.
:snowflake: API server now serves the history only on demand `?history=yes` -> breaking change
:snowflake: We store `StateEvent` on disk instead of `ServerOutput`
:snowflake: Our projections now work with `StateChanged` type.
:snowflake: Some of the messages like `Greetings`, `PostTxOnChainFailed` are served directly from the api without `time` and `seq` fields so they need to be decoded separately (more work on the client side)

 

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ch1bo committed:
Draft key changes to the API server becoming an event sink

v0d1ch committed:
Wire things together

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use eventSource instead of persistence in the Server

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add time to StateEvent

v0d1ch committed:
Fix Server tests

v0d1ch committed:
Map more fields between ServerOutput and Outcome

v0d1ch committed:
BehaviorSpec

v0d1ch committed:
NodeSpec

v0d1ch committed:
Golden files

v0d1ch committed:
HeadLogicSpec, removed persistence and renamed Api to ApiSink, re-use time from StateChanged

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
BehaviorSpec trace test and inline projections update in Server

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add persistenc to the api sink so we can source saved events

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Clean up HeadLogic

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Format, cleanup

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Remove history TVar and set default to not serve history

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo commented on hydra-node/golden/StateChanged/CommitRecorded.json:
This file is massive and there is no reason why. We should be using `size = 1` for these generators as we are not interested in big `UTxO` sets (for example), but the basic structure of the `CommitRecorded` value.

Should: reduce the diff by using smaller quickcheck sizes when generating golden files

ch1bo commented on hydra-node/golden/StateEvent (Tx ConwayEra).faulty.json:
Must: not commit faulty golden files files

ch1bo commented on hydra-node/golden/StateChanged/GetUTxOResponse.json:
Should: not map this to a `StateChanged` / `StateEvent`

ch1bo commented on hydra-node/golden/StateChanged/Greetings.json:
Must: not map this to a `StateChanged` / `StateEvent`

Greetings has *nothing* to do with the head logic and is a 100% procedural response created by the API server for connecting clients.

ch1bo commented on hydra-node/golden/StateChanged/InvalidInput.json:
Must: not map this to a `StateChanged` / `StateEvent`

This has nothing to do with the HeadLogic and was a client input that could not get parsed.

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -1404,7 +1404,7 @@ components:
         headId:
           $ref: "api.yaml#/components/schemas/HeadId"
-        theDeposit:
+        depositTxId:
```
Must: mention this as a breaking change in the `CHANGELOG.md`

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -98,19 +99,22 @@ withAPIServer config env party persistence tracer chain pparams serverOutputFilt
-
-    history <- newTVarIO loadedHistory
     -- NOTE: we need to reverse the list because we store history in a reversed
     -- list in memory but in order on disk
```
Must: remove irrelevant comments when changing code

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -98,19 +99,22 @@ withAPIServer config env party persistence tracer chain pparams serverOutputFilt
+                    update pendingDepositsP output
+            )
+          .| mapWhileC mkTimeServerOutputFromStateEvent
           .| sinkList
```
Must: not load the history into memory (using `sinkList`)!

This is the whole point of this issue/PR.

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -137,7 +141,7 @@ withAPIServer config env party persistence tracer chain pparams serverOutputFilt
             Server
               { sendOutput = \output -> do
-                  timedOutput <- appendToHistory history output
+                  timedOutput <- mkTimedOutput history output
```
Must: not have two sources of `seq`

This function and `mkTimeServerOutputFromStateEvent` result in potentially different sequences of integers, which will violate the invariant of always increasing `seq` numbers.

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -98,19 +99,22 @@ withAPIServer config env party persistence tracer chain pparams serverOutputFilt
+                    update headIdP output
+                    update pendingDepositsP output
+            )
+          .| mapWhileC mkTimeServerOutputFromStateEvent
```
Should: move this up and not call `mkTimeServerOutputFromStateEvent` twice

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -137,7 +141,7 @@ withAPIServer config env party persistence tracer chain pparams serverOutputFilt
           action $
             Server
               { sendOutput = \output -> do
```
Should: accept a `StateEvent` or just return an `EventSink`

This will make the caller provide the sequence number (event id) and `time`stamp already.

v0d1ch replied:
I opted in to accept a `StateEvent` type here.

ch1bo commented on hydra-node/src/Hydra/API/WSServer.hs:
```diff
@@ -75,7 +74,7 @@ wsApp party tracer history callback headStatusP headIdP snapshotUtxoP responseCh
 
   -- api client can decide if they want to see the past history of server outputs
-  unless (shouldNotServeHistory queryParams) $
+  when (shouldServeHistory queryParams) $
```
I never understood why we make this breaking change in part of this (constant memory) refactoring?

Whether we serve the history by default or not does not change the semantics of loading the history from the event stream (on disk).

v0d1ch replied:
It was easy to do and it was mentioned as one of the items to do in the api server refactoring. It would be a small PR anyway so I included it here.

ch1bo commented on hydra-node/src/Hydra/API/WSServer.hs:
```diff
@@ -53,7 +52,7 @@ wsApp ::
   Party ->
   Tracer IO APIServerLog ->
-  TVar [TimedServerOutput tx] ->
+  [TimedServerOutput tx] ->
```
Must: not load the whole server output into memory

ch1bo commented on hydra-node/src/Hydra/API/WSServer.hs:
```diff
@@ -161,14 +160,13 @@ wsApp party tracer history callback headStatusP headIdP snapshotUtxoP responseCh
         time <- getCurrentTime
-        seq <- atomically $ nextSequenceNumber history
+        let seq = nextSequenceNumber history
         let timedOutput = TimedServerOutput{output = InvalidInput @tx e clientInput, time, seq}
```
This is tricky now as we cannot increment the sequence number because it is ultimately determined by the `Hydra.Node` which creates the `StateEvent` stream.

I see two options:

1. We define `seq` to be always `>=` than and use the same as we last saw before sending this error response. We need to keep track of last seen sequence numbers as they flow through the `chan` in this case.

2. We do not include a `seq` with (error) responses like this.

v0d1ch replied:
Now `nextSequenceNumber` runs a conduit to determine the correct seq number.

v0d1ch replied:
We opted in for `seq` removal in our mob session. Now the server outputs `Greetings` `InvalidInput` etc just as normal server outputs, no time or seq fields anymore.

ch1bo commented on hydra-node/src/Hydra/Events/ApiSink.hs:
```diff
@@ -0,0 +1,43 @@
+  pure EventSink{putEvent}
+ where
+  PersistenceIncremental{append, source} = persistence
+  Server{sendOutput} = server
```
Must: not need this

This is not needed as every event sink in `HydraNode.eventSinks` gets all events delivered, once on load and later for each event.

Hence we can create an `EventSink` directly by interpret using a `Server` (or remove `Server` altogether).

v0d1ch replied:
I don't understand this comment. I don't see how I could remove `Server` from here (I already removed the persistence and need to think about how to make certain tests pass because of this)

ch1bo commented on hydra-node/src/Hydra/Events/FileBased.hs:
```diff
@@ -27,11 +27,14 @@ import Hydra.Persistence (PersistenceIncremental (..))
 eventPairFromPersistenceIncremental PersistenceIncremental{append, source} = do
   eventIdV <- newTVarIO Nothing
+  -- TODO: All legacy events will have the same associated time. How to work around this?
+  now <- getCurrentTime
```
This is not only an issue for legacy events, but also for "new" events as we are _adding_ the `time` field in this PR. In fact, we can and should drop the `Legacy` parsing as we have released (major) versions of the `hydra-node` which required wiping of the persistence dir (double-check please)

Should: get current time and add it while parsing (move `getCurrentTime` to conduit)
Should: drop the `Legacy` mode (and mention it in CHANGELOG)

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -206,14 +203,10 @@ onInitialChainCommitTx ::
-      ( notifyClient
-          : [postCollectCom | canCollectCom]
-      )
+      ([postCollectCom | canCollectCom])
```
Hlint: Redundant parenthesis

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -322,32 +313,31 @@ onOpenNetworkReqTx env ledger st ttl tx =
-          -- Spec: if ŝ = ̅S.s ∧ leader(̅S.s + 1) = i
-          --         multicast (reqSn, v, ̅S.s + 1, T̂ , 𝑈𝛼, txω )
-          & maybeRequestSnapshot (confirmedSn + 1)
+      -- (cause (ClientEffect $ ServerOutput.TxValid headId (txId tx) tx) <>) $
```
Must: remove commented code

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -322,32 +313,31 @@ onOpenNetworkReqTx env ledger st ttl tx =
+  -- because of network latency when receiving tx1. The leader,
+  -- however, saw both as valid and requests a snapshot including
+  -- both. This is a valid request and if we would have removed tx2
+  -- from allTxs, we would make the head stuck.
```
Should move the comment between `otherwise` and `newState TxInvalid` to avoid it going out of context

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1168,44 +1118,36 @@ onClosedChainContestTx ::
+                    }
+              }
+    | snapshotNumber > number (getSnapshot confirmedSnapshot) ->
+        noop
```
Must: not change semantics!

This likely is a bug. Before we always produced the `HeadContested` event and now this branch of the if does not produce anything!

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1288,7 +1229,7 @@ update env ledger st ev = case (st, ev) of
     | otherwise -> Error NotOurHead{ourHeadId, otherHeadId = headId}
   (Initial InitialState{committed, headId}, ClientInput GetUTxO) ->
-    cause (ClientEffect . ServerOutput.GetUTxOResponse headId $ fold committed)
+    newState (GetUTxOResponse headId $ fold committed)
```
Should: drop this query response as it is now infecting the `StateEvent` stream.

I would strongly argue for this not be needed anymore as we have the `GET /snapshot/utxo` query. _If_ we want to keep this backwards compatible, the websocket server should get the data the same way as the `GET /snapshot/utxo` handler in the http server does.

In any case, we will not have a `GetUTxOResponse` state event.

v0d1ch replied:
I think so too. Removed.

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1308,10 +1249,10 @@ update env ledger st ev = case (st, ev) of
-     in cause (ClientEffect . ServerOutput.GetUTxOResponse headId $ getField @"utxo" snapshot' <> fromMaybe mempty (getField @"utxoToCommit" snapshot'))
+     in newState (GetUTxOResponse headId $ getField @"utxo" snapshot' <> fromMaybe mempty (getField @"utxoToCommit" snapshot'))
+  -- -- TODO: Is it really intuitive that we respond from the confirmed ledger if
+  -- -- transactions are validated against the seen ledger?
```
Same as above and please also remove this TODO (it got out of context!)

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1349,8 +1290,8 @@ update env ledger st ev = case (st, ev) of
-        newState HeadIsReadyToFanout
-          <> cause (ClientEffect $ ServerOutput.ReadyToFanout headId)
+        newState HeadIsReadyToFanout{headId}
+  -- <> cause (ClientEffect $ ServerOutput.ReadyToFanout headId)
```
Must: remove commented code

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1364,9 +1305,9 @@ update env ledger st ev = case (st, ev) of
+    newState PostTxOnChainFailed{postChainTx, postTxError}
   (_, ClientInput{clientInput}) ->
-    cause . ClientEffect $ ServerOutput.CommandFailed clientInput st
+    newState CommandFailed{clientInput, state = st}
```
Event sourcing question (@ffakenz you seemed to have experience in this IIRC):

Does this look like a good event in an event source world? This implies that `CommandFailed` gets persisted and **re-aggregated** when (re-)starting the `hydra-node`.

v0d1ch replied:
We should remove `CommandFailed` since we can just decide to use `Error` catch pattern here and remove it from `StateChanged`

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1364,9 +1305,9 @@ update env ledger st ev = case (st, ev) of
-    cause . ClientEffect $ ServerOutput.CommandFailed clientInput st
+    newState CommandFailed{clientInput, state = st}
   _ ->
     Error $ UnhandledInput ev st
```
In fact.. do we need the case above as we have this `Error` and it a `ClientInput` not being handled is rather a "programmer error" than a "user error"? i.e. the user of the API should not be at fault and can't change their query if we forgot to handle it in a given state?

v0d1ch replied:
I removed the `CommandFailed` as well as this case match.

ch1bo commented on hydra-node/src/Hydra/Logging/Monitoring.hs:
```diff
@@ -96,16 +92,6 @@ monitor transactionsMap metricsMap = \case
-          atomically $ modifyTVar' transactionsMap $ Map.delete (txId tx)
-          histo "hydra_head_tx_confirmation_time_ms" (diffTime t start)
-        Nothing -> pure ()
-    tickN "hydra_head_confirmed_tx" (length $ confirmed snapshot)
```
Must: not break prometheus metrics

This is dropping metrics from the API and not scope of this PR? Or do you disagree?

v0d1ch replied:
We don't have `ClientEffect` anymore. This is the only reason why this branch was removed.

v0d1ch replied:
We could pattern match on `LogicOutcome` here.

ch1bo commented on hydra-node/src/Hydra/Node/Run.hs:
```diff
@@ -91,15 +91,16 @@ run opts = do
-              connect chain network server wetHydraNode
+              connect chain network wetHydraNode
                 >>= runHydraNode
+                . addApiEventSink apiServerSink
```
Could: Move this up so it's clear that we run the hydra node with the api sink added.

```suggestion
              connect chain network wetHydraNode
                <&> addApiEventSink apiServerSink
                >>= runHydraNode
```

ch1bo commented on hydra-node/src/Hydra/Node/Run.hs:
```diff
@@ -91,15 +91,16 @@ run opts = do
             -- Network
             let networkConfiguration = NetworkConfiguration{persistenceDir, signingKey, otherParties, host, port, peers, nodeId}
             withNetwork tracer networkConfiguration (wireNetworkInput wetHydraNode) $ \network -> do
+              apiServerSink <- wireApiEvents incrementalPersistence server
```
Must: not use the persistence to create the api server sink

This is not needed. See my other comment further above. This should be possible with something like:
```suggestion
              let apiServerSink = wireApiEvents server
```

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -837,17 +841,20 @@ spec = parallel $ do
-      logs `shouldContain` [BeginEffect alice 2 0 (ClientEffect $ HeadIsInitializing testHeadId $ fromList [alice])]
-      logs `shouldContain` [EndEffect alice 2 0]
+    logs `shouldSatisfy` \case
+      as -> expectedOutcome `elem` as
```
Should: just use `shouldContain`?
```suggestion
    logs `shouldContain` [expectedOutcome]
```

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -110,9 +111,11 @@ spec =
-        update bobEnv ledger s0 reqTx `hasEffectSatisfying` \case
-          ClientEffect TxInvalid{transaction} -> transaction == tx
+        update bobEnv ledger s0 reqTx `shouldSatisfy` \case
+          Continue as _ -> expectedStateChange `elem` as
```
Should: not complicate test assertions

We do not want to assert the full `ValidationError`. Instead please translate the `hasEffectSatisfying` helper to a `hasStateChangeSatisfying` (or so) that it can stay:

```
update bobEnv ledger s0 reqTx `hasStateChangeSatisfying` \case
          TxInvalid{transaction} -> transaction == tx
```

ch1bo replied:
This applies to all changes in this file.

ch1bo commented on hydra-node/test/Hydra/Logging/MonitoringSpec.hs:
```diff
@@ -28,7 +27,8 @@ spec =
         threadDelay 0.1
-        traceWith tracer (Node $ BeginEffect alice 0 0 (ClientEffect (SnapshotConfirmed testHeadId (testSnapshot 1 1 [tx2, tx1] (utxoRefs [1])) mempty)))
+        -- FIXME: This is not working, the message is not being logged
+        -- traceWith tracer (Node $ BeginEffect alice 0 0 (ClientEffect (SnapshotConfirmed testHeadId (testSnapshot 1 1 [tx2, tx1] (utxoRefs [1])) mempty)))
```
Must: fix the metrics (see other comment)

ch1bo reviewed (changes requested):
The scope is more-or-less fine, but we can avoid some breaking changes (if we want).

Biggest issue is that this still loads all events into memory and the api server can be represented as an `EventSink` much more easy.

The diff contains at least one bug and a lot of commented code, outdated comments and at lest one FIXME.

v0d1ch committed:
WIP: Pr review comments

- 'sendOutput' now serves `StateEvent`.
- We do not load complete history with `sinkList`.
- Removed `GetUTxO`.
- Remove persistence from the ApiSink.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch commented on CHANGELOG.md:
```diff
@@ -23,6 +23,10 @@ changes.
+- **BREAKING**
+  - Remove `GetUTxO` client input and corresponding `GetUTxOResponse`. There is already a way to query the `UTxO` in the Head with `GET /snapshot/utxo` query.
+  - Renamed 'CommitFinalized' field 'theDeposit' to 'depositTxId'.
+
```
Add this to the changelog: Hydra node API server is used as event sink and should perform better (using bounded memory)

v0d1ch replied:
Add note about history serving not being the default any more.

v0d1ch commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -98,21 +99,18 @@ withAPIServer config env party persistence tracer chain pparams serverOutputFilt
-    -- NOTE: we need to reverse the list because we store history in a reversed
-    -- list in memory but in order on disk
+    let history =
+          sourceEvents
```
We should still run this conduit here (to prime the projections) and then pass the conduit to websocket server.

ch1bo commented on hydra-node/src/Hydra/Node/Run.hs:
```diff
@@ -91,15 +90,16 @@ run opts = do
-              connect chain network server wetHydraNode
-                >>= runHydraNode
+              connect chain network wetHydraNode
+                <&> addApiEventSink apiServerSink
```
```suggestion
                <&> addEventSink EvenSink{putEvent = sendOutput server}
```

v0d1ch commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -347,7 +347,6 @@ processEffects node tracer inputId effects = do
   processEffect (effect, effectId) = do
     traceWith tracer $ BeginEffect party inputId effectId effect
     case effect of
-      ClientEffect i -> sendOutput server i
```
We should still keep the `ClientEffect` for the purpose of sending some messages to the client that we don't want persisted. Also keep the `GetUTxOResponse`.

v0d1ch replied:
I ended up removing it since it requires `ServerOutput` and in turn we need to accept the same type in `Server{sendOutput}` and now we accept `StateChanged`. We could keep track of `seq` numbers and do it this way but our original plan was to actually remove `ClientEffect` so I ended up doing that.

v0d1ch commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -1395,7 +1395,7 @@ components:
       required:
```
We remove sequence numbers on messages coming from the api server that do not come from `StateEvent`. For example: `Greetings`, `InvalidInput`, `CommandFailed`, `PostTxOnChainFailed` etc.

v0d1ch commented on hydra-node/src/Hydra/API/WSServer.hs:
```diff
@@ -161,14 +162,13 @@ wsApp party tracer history callback headStatusP headIdP snapshotUtxoP responseCh
         time <- getCurrentTime
-        seq <- atomically $ nextSequenceNumber history
+        seq <- nextSequenceNumber history
         let timedOutput = TimedServerOutput{output = InvalidInput @tx e clientInput, time, seq}
```
NOTE: we don't need to output `seq` here at all.

v0d1ch commented on hydra-node/src/Hydra/API/WSServer.hs:
```diff
@@ -87,7 +88,7 @@ wsApp party tracer history callback headStatusP headIdP snapshotUtxoP responseCh
   -- important to make sure the latest configured 'party' is reaching the
   -- client.
   forwardGreetingOnly con = do
-    seq <- atomically $ nextSequenceNumber history
```
NOTE: We don't need to output `TimeServerOutput` here. We could just send out `Greetings` type.

v0d1ch committed:
Updated with review suggestions

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Query HeadState to replace GetUTxO input

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Query the node state in Model tests

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Replace GetUTxO in hydra-cluster with "snapshot/utxo" request

Cleanup

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Remove GetUTxO client input and fix hydraw

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Fix hydra-tui tests

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Always set `history=yes` for tui and cluster clients

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use LogicOutcome instead of ClientEffect for monitoring tests

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
HeadLogic spec uses better helpers now

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Remove seq and time from outputs that do not map to StateChanged

v0d1ch committed:
Small refactor for BehaviorSpec

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Remove CommandFailed completely

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Fix flakiness in Model tests by retrying

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Project over StateChanged, don't use mapWhileC

Problem was that our mapping is not complete and mapWhileC is stopping
if it encounteres partial mapping from StateEvent to ServerOutput. This
was causing our history to not work properly.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Introduce registerThread to PersistenceIncremental

Now append is checked against the correct thread id but we allow
source to work for any client.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Introduce MinimumSized

Still need to find out where to control output for golden/StateChanged

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Remove registerThread

Directly register thread id when using append (if it doesn't exist)
since it is hard to registerThread id inside of a running process and
there is no real need for additional field (initially I thought it might
be good to keep track of different thread Ids for appending but in
reality we need to write from only one thread).

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Remove Legacy handling of events

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo commented on hydra-node/golden/StateChanged/Greetings.json:
Must: not have stale golden files

ch1bo commented on hydra-node/golden/StateChanged/TransactionAppliedToLocalUTxO.json:
Should: use `MinimumSized` on this (and other) updated golden file

v0d1ch replied:
How do I do that? I am not sure which golden tests were adding these files.

ch1bo replied:
Change [this](https://github.com/cardano-scaling/hydra/blob/8072ca58c573e1359935f9569fffd27c17604997/hydra-node/test/Hydra/Events/FileBasedSpec.hs#L31) to `Proxy (MinimumSized (StateChanged Tx))`

ch1bo commented on hydra-node/src/Hydra/Events/ApiSink.hs:
```diff
@@ -0,0 +1,7 @@
+import Hydra.Node (HydraNode (..))
+
+addEventSink :: EventSink (StateEvent tx) m -> HydraNode tx m -> HydraNode tx m
+addEventSink sink node = node{eventSinks = sink : eventSinks node}
```
Should: move this to `Hydra.Node` or `Hydra.Events`

I think this is too little code for a module.

ch1bo commented on hydra-node/src/Hydra/Events/FileBased.hs:
```diff
@@ -9,26 +9,18 @@ import Conduit (mapMC, (.|))
 -- A new implementation for an 'EventSource' with a compatible 'EventSink' could
 -- be defined more generically with constraints:
 --
 -- (ToJSON e, FromJSON e, HasEventId) e => (EventSource e m, EventSink e m)
```
Should: follow this comment and remove it

This is now possible.

v0d1ch replied:
Remove what? I don't get it.

ch1bo replied:
Remove the haddock

> -- A new implementation for an 'EventSource' with a compatible 'EventSink' could
-- be defined more generically with constraints:
--
-- (ToJSON e, FromJSON e, HasEventId) e => (EventSource e m, EventSink e m)

and switch to the signature as proposed.

ch1bo commented on hydra-node/src/Hydra/HeadLogic/Outcome.hs:
```diff
@@ -84,49 +100,63 @@ data StateChanged tx
   arbitrary = arbitrary >>= genStateChanged
 
-instance (ArbitraryIsTx tx, IsChainState tx) => ToADTArbitrary (StateChanged tx)
+instance (ArbitraryIsTx tx, IsChainState tx) => ToADTArbitrary (MinimumSized (StateChanged tx))
```
Should: not need this here, but define the instance on `MinimumSized`

```suggestion
instance (ArbitraryIsTx tx, IsChainState tx) => ToADTArbitrary (StateChanged tx)
```

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -866,14 +841,7 @@ onOpenNetworkReqDec env ledger ttl openState decommitTx =
-              , utxoToDecommit = decommitUTxO
-              }
-        )
+      <> newState DecommitRequested{headId, decommitTx = decommitTx, utxoToDecommit = decommitUTxO}
```
Must: not add `DecommitRequested` as we already have `DecommitRecorded`

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1168,44 +1116,35 @@ onClosedChainContestTx ::
+    | snapshotNumber > number (getSnapshot confirmedSnapshot) ->
+        newState HeadContested{headId, chainState = newChainState, contestationDeadline, snapshotNumber}
+    -- TODO: A more recent snapshot number was succesfully contested, we will
+    -- not be able to fanout! We might want to communicate that to the client!
```
Must: not shuffle comments

```suggestion
        -- TODO: A more recent snapshot number was succesfully contested, we will
        -- not be able to fanout! We might want to communicate that to the client!
        newState HeadContested{headId, chainState = newChainState, contestationDeadline, snapshotNumber}

```

ch1bo commented on hydra-node/src/Hydra/HeadLogic/State.hs:
```diff
@@ -68,6 +68,13 @@ getHeadParameters = \case
+getHeadUTxO = \case
+  Open OpenState{coordinatedHeadState = CoordinatedHeadState{localUTxO}} -> Just localUTxO
+  Initial InitialState{committed} -> Just $ foldl' (<>) mempty $ toList committed
+  _ -> Nothing
```
Should: move this to test code

It seems this is only used from tests?

ch1bo commented on hydra-node/src/Hydra/Node.hs:
```diff
@@ -319,11 +318,12 @@ processNextInput HydraNode{nodeState, ledger, env} e =
+  events <- forM stateChanges $ \stateChanged -> do
+    time <- getCurrentTime
+    eventId <- atomically getNextEventId
+    pure StateEvent{eventId, stateChanged, time}
```
:+1:

ch1bo commented on hydra-node/src/Hydra/Node/Run.hs:
```diff
@@ -91,15 +92,15 @@ run opts = do
-              connect chain network server wetHydraNode
-                >>= runHydraNode
+              connect chain network wetHydraNode
+                <&> addEventSink EventSink{putEvent = sendOutput server}
```
:+1:

ch1bo commented on hydra-node/src/Hydra/Persistence.hs:
```diff
@@ -91,22 +92,25 @@ createPersistenceIncremental fp = do
+            Just tid -> pure tid
+          appendTid <- myThreadId
+          when (appendTid /= tid) $
+            throwIO (IncorrectAccessException $ "Trying to append to persisted data in " <> fp <> " from different thread")
```
Must: update haddocks as this is changing semantics of "thread protection"

There is also a test in `PersistenceSpec` which probably needs an update because of this.

While it makes sense that we are sourcing from multiple threads now, I'm not sure if this is safe. This was the original reason why we started to limit access to the same thread: https://github.com/cardano-scaling/hydra/issues/1202

I think we should record somewhere the requirement that the `EventSource` and `EventSink` need to be able to be used from different threads (probably in `Hydra.Events.FileBased.hs`) and that this may not be safe currently. Later, we could explore better means to write to disk than line-by-line using JSON .. e.g. a simple embedded db like sqlite or rocksdb.

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -220,39 +227,46 @@ spec =
+                        Left{} -> failure $ "Failed to decode messages:\n" <> show received
+                        Right _ -> pure []
+                    Right (timedOutputs :: [TimedServerOutput SimpleTx]) -> pure timedOutputs
+                  seq <$> results `shouldSatisfy` isContinuous
```
How about just collecting all `key "seq"` from the `received`?

Something like `let results = received ^.. key "seq"`

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -220,39 +227,46 @@ spec =
+        -- XXX: here we manually update eventId's to be sequential since it is easier than in the arbitrary instance.
+        -- API Server should keep these intact and just produce corresponding sequence numbers. In real life persistence
+        -- would handle this behavior anyway.
+        -- let outputs = zipWith (\stateEvent eventId -> stateEvent{eventId}) outputs' [1 ..]
```
Must: remove commented code

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -261,50 +275,47 @@ spec =
+          snapShotConfirmedMsg@Outcome.SnapshotConfirmed{snapshot = Snapshot{utxo, utxoToCommit}} <-
+            generateSnapshot
+          headIsInitializing :: Outcome.StateChanged SimpleTx <- generate $ Outcome.HeadInitialized <$> arbitrary <*> arbitrary <*> arbitrary <*> arbitrary <*> arbitrary
+          let expectedUtxos = toJSON $ utxo <> fromMaybe mempty utxoToCommit
```
Why is this  now including `utxoToCommit`?

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -202,3 +191,95 @@ setupServerNotification :: IO (NotifyServerRunning, WaitForServer)
+-- | Projection of latest confirmed snapshot UTxO.
+projectSnapshotUtxo :: IsTx tx => Maybe (UTxOType tx) -> StateChanged.StateChanged tx -> Maybe (UTxOType tx)
+projectSnapshotUtxo snapshotUtxo = \case
+  StateChanged.SnapshotConfirmed _ snapshot _ -> Just $ Tx.utxo snapshot <> fromMaybe mempty (Tx.utxoToCommit snapshot)
```
Why did this change?

It was not including `utxoToCommit` before:
```haskell
-- | Projection of latest confirmed snapshot UTxO.
projectSnapshotUtxo :: Maybe (UTxOType tx) -> ServerOutput tx -> Maybe (UTxOType tx)
projectSnapshotUtxo snapshotUtxo = \case
  SnapshotConfirmed _ snapshot _ -> Just $ Tx.utxo snapshot
  HeadIsOpen _ utxos -> Just utxos
  _other -> snapshotUtxo
```

ch1bo commented on hydra-node/src/Hydra/API/WSServer.hs:
```diff
@@ -160,25 +154,17 @@ wsApp party tracer history callback headStatusP headIdP snapshotUtxoP responseCh
+    -- NOTE: we need to reverse the list because we store history in a reversed order on disk
+    let hist = reverse hist'
     let encodeAndReverse xs serverOutput = Aeson.encode serverOutput : xs
     sendTextDatas con $ foldl' encodeAndReverse [] hist
```
Must: not load the whole history into memory!

You probably want to invoke `sendTextData` from within the conduit.

I also don't think we need to reverse the history here?

ch1bo replied:
Oh.. it reverses the list twice!

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -104,90 +106,93 @@ spec =
             withFreePort $ \port ->
-              withTestAPIServer port alice mockPersistence tracer $ \Server{sendOutput} -> do
+              withTestAPIServer port alice (mockSource history) tracer $ \Server{sendOutput} -> do
                 let sendFromApiServer = sendOutput
```
This is a pointless `let`?

ch1bo commented on hydra-node/test/Hydra/API/ServerSpec.hs:
```diff
@@ -198,18 +203,20 @@ spec =
                     Left{} -> failure $ "Failed to decode messages:\n" <> show received
                     Right timedOutputs' -> do
-                      (output <$> timedOutputs') `shouldBe` [notHistoryMessage]
+                      (output <$> timedOutputs') `shouldBe` [fromMaybe (error "failed to convert in mapStateChangedToServerOutput") (mapStateChangedToServerOutput $ stateChanged notHistoryMessage)]
```
Isn't this flaky as it `notHistoryMessage` may very well be an arbitrary `StateEvent`?

Instead of working around on the assertion, I would suggest to generate more relevant data. i.e. limit the `Gen (StateEvent tx)` to only produce values which are sent out on the api server. For example

```haskell
genStateEventForApi :: Gen (StateEvent tx)
genStateEventForApi =
  arbitrary `suchThat` (isJust . mkTimedServerOutputFromStateEvent)
```

This comment also applies to other cases in this file.

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -172,10 +177,14 @@ spec = parallel $ do
+              waitUntilMatch [n1, n2] $ \case
+                SnapshotConfirmed{snapshot = Snapshot{number, confirmed}} ->
+                  number == 1 && confirmed == [aValidTx 42]
+                _ -> False
```
So is there no `CommandFailed` anymore if we try to do an `Abort` here?

IMO Having a response for failed commands is important and this refactoring PR should not degrade user experience.

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -372,9 +380,8 @@ spec = parallel $ do
-
-                waitUntil [n1] $ GetUTxOResponse testHeadId (utxoRefs [2, 42])
+                headUTxO <- getHeadUTxO <$> queryState n1
+                fromMaybe mempty headUTxO `shouldBe` utxoRefs [2, 42]
```
Should: drop this test.

This test is about `GetUTxO`, which you dropped.

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -202,9 +211,8 @@ spec = parallel $ do
-
-              waitUntil [n2] $ GetUTxOResponse testHeadId (utxoRefs [1])
+              headUTxO <- getHeadUTxO <$> queryState n2
+              fromMaybe mempty headUTxO `shouldBe` utxoRefs [1]
```
Could: just don't do this assertion.

The `waitUntil [n2] $ Committed ...` is good enough.

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -837,17 +833,19 @@ spec = parallel $ do
 
-      logs `shouldContain` [BeginEffect alice 2 0 (ClientEffect $ HeadIsInitializing testHeadId $ fromList [alice])]
-      logs `shouldContain` [EndEffect alice 2 0]
+    logs `shouldContain` [expectedOutcome]
```
Must: re-indent (not change) this

I don't see why this was moved?

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -655,11 +654,8 @@ spec = parallel $ do
-                      GetUTxOResponse{headId, utxo} -> headId == testHeadId && not (member 42 utxo)
-                      _ -> False
+                  headUTxO <- getHeadUTxO <$> queryState n1
+                  fromMaybe mempty headUTxO `shouldBe` utxoRefs [2]
```
I think none of the `GetUTxO` replacements is needed in these tests.

ch1bo commented on hydra-prelude/src/Hydra/Prelude.hs:
```diff
@@ -214,6 +215,16 @@ newtype ReasonablySized a = ReasonablySized a
+minimumSized :: Gen a -> Gen a
+minimumSized = scale (const 1)
+
+-- | A QuickCheck modifier that allows scaled down values.
```
Might want to say that we are using `size = 1`?

ch1bo commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -938,20 +935,26 @@ waitForUTxOToSpend utxo key value node = go 100
+  TestHydraClient tx m ->
+  m (UTxOType tx)
+headUTxO node = do
+  threadDelay 1
```
Why is here a delay?

v0d1ch replied:
Otherwise the test becomes _flaky_ and just consumes a lot of memory and never finishes. I didn't find the reason for this yet and delaying in `waitForUTxOToSpend` does the trick, but yes, we should investigate.

ch1bo replied:
Maybe you are calling `headUTxO` in a loop somewhere which becomes a busy loop (without delays between retries) if you drop this?

Should: not delay here, but in a loop if there is one

ch1bo commented on hydra-node/test/Hydra/ModelSpec.hs:
```diff
@@ -216,10 +216,12 @@ assertBalancesInOpenHeadAreConsistent ::
     Open{offChainState = OffChainState{confirmedUTxO}} -> do
-      utxo <- run $ getUTxO node
+      utxo <- run $ lift $ headUTxO node
+      run $ lift $ threadDelay 1
```
Why are delays added here and above?

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -725,7 +717,6 @@ components:
         - $ref: "api.yaml#/components/schemas/GetUTxOResponse"
```
Must: remove all removed things from api schema

At least `GetUTxOResponse` here is not a thing anymore

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -320,7 +319,7 @@ components:
       title: Init
       description: |
-        Initializes a new Head. This command is a no-op when a Head is already open and the server will output an `CommandFailed` message should this happen.
+        Initializes a new Head. This command is a no-op when a Head is already open.
```
Must: not change these semantics.. the error message is helpful!

ch1bo commented on hydra-tui/src/Hydra/Client.hs:
```diff
@@ -84,7 +86,11 @@ withClient Options{hydraNodeHost = Host{hostname, port}, cardanoSigningKey, card
+        -- TODO: same double decoding problem here
+        case eitherDecodeStrict msg of
+          Right output -> callback $ UpdateDirect output
+          Left err -> throwIO $ ClientJSONDecodeError err msg
```
What is the difference between `Update` and `UpdateDirect`?

Why don't you just decode all the possible messages here? I assume it's about `TimedServerOutput` not unifying all possible messages?

I would suggest the following:
- Create a type that enumerates all things we can receive, e.g. `data HydraMessage = ReceivedServerOutput (ServerOutput tx) | ReceivedGreetings Greetings | InvalidInput | ...`
- Define the `FromJSON` instance using alternative, e.g. `parseJSON v = (ReceivedServerOutput <$> parseJSON v) <|> (ReceivedGreetings <$> parseJSON v) <|> ...`
- Use that type in `Update (HydraMessage tx)`

ch1bo reviewed (changes requested):
This is still loading all events into memory when we open a websocket connection and I'm not sure I like dropping `CommandFailed`

v0d1ch committed:
Rebase fixes

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Stream history from websocket server instead of loading it to memory

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
PR review: Bring back ClientInput, CommandFailed etc.

v0d1ch committed:
Introduce HydraMessage

This is a message containing any possible type we want to send our from
our web server (Server{sendMessage}). This is needed in order to get
back `ClientEffect` that sends out `ServerOutput` but also keep
`StateEvent` as the persisted type and outputted from the api server.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use server in NodeSpec to record messages

Fix the api.yaml error

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Remove network messages from stored StateEvent

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Draft expected message behavior in the API

Just the first round to compile hydra-node, still need to think what
is the best design for Behavior, Node, HeadLogic spec.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add missing golden files for network messages

Remove unused type aliases

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
ServerSpec - Introduce AllPosibleAPIMessages for easier parsing

No need to server utxoToCommit in the relevant HeadState utxo projection
Make sure Greetings omits Nothing fields

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use AllPosibleAPIMessages in the tui

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use AllPosibleAPIMessages in Behavior/Model

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo committed:
GET /snapshot/utxo should include committed utxo

As we are projecting the confirmed snapshot, this means those utxo are
available for spending.

v0d1ch committed:
Not use AllPossibleAPIMessages in BehaviorSpec

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo committed:
Not use AllPossibleAPIMessages in ServerSpec

ch1bo committed:
Fix Model and move AllPossibleAPIMessages

ch1bo committed:
Fix api and logs schemas

ch1bo committed:
Fix ServerSpec tests dealing with history

v0d1ch committed:
Fix ServerSpec tests that rely on update Head UTxO projection

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Fix ServerSpec tests that rely on ServerOutput api schema

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Fix NodeSpec re-hydration test

It was shortcutting on failed mapping StateEvent -> ServerOutput

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Fix BehaviorSpec trace test

It seems like waitForMessage was not needed in the end

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo committed:
Fix NodeSpec

ch1bo committed:
Fix BehaviorSpec

v0d1ch committed:
Fix e2e spec related to the now non existing server-output file

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Remove unused Arbitrary instances

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Remove ApiSink module

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo committed:
Remove redundant /snapshot/utxo getters

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -443,7 +443,6 @@ spec = around (showLogsOnFailure "EndToEndSpec") $ do
               -- not clearing the whole persistence dir as we would not be able
               -- to re-connect to the L2 network.
               removeFile $ tmp </> "state-" <> show bobNodeId </> "state"
-              removeFile $ tmp </> "state-" <> show bobNodeId </> "server-output"
```
:heart:

ch1bo committed:
Test server output spec conformance again

ch1bo committed:
Only use mkTimedServerOutputFromStateEvent

This removes a duplication of creating the timed server output value
from a state event.

ch1bo committed:
Unshuffle comments

v0d1ch committed:
Use HasEventId for file persistence to make it more abstract

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo committed:
Add ToADTArbitrary instance to MinimumSized

ch1bo committed:
Regenerate StatChanged golden files with minimum size

ch1bo committed:
Drop redundant ServerOutput golden files

Also use MinimumSized already, but not re-generate (as not needed right now)

ch1bo committed:
Add golden files for client input using ToADTArbitrary

ch1bo committed:
Supersede validate-api.js with our haskell-based tests

We realized that we are basically doing the same thing in prop_specIsComplete.

ch1bo committed:
Small HeadLogicSpec refactor

ch1bo reviewed (approved):
The model thread delays are weird.. let's look at them another time.

v0d1ch committed:
Fix HeadLogic test related to network messages

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>
